design/userdata-merging.md

```diff
@@ -0,0 +1,234 @@
+      * [Design Details](#design-details)
+         * [Work Items](#work-items)
+         * [Dependencies](#dependencies)
+         * [Test Plan](#test-plan)
+         * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+         * [Version Skew Strategy](#version-skew-strategy)
+      * [Drawbacks](#drawbacks-optional)
+      * [Alternatives](#alternatives-optional)
+      * [References](#references)
+
+<!-- Added by: stack, at: 2019-02-15T11:41-05:00 -->
+
+<!--te-->
+
+## Summary
+
+With v1alpha2, Cluster-API goal is to split the bootstrap provider from the
+infrastructure provider. However, in the case of baremetal, such a split is not
+straightforward. Some configuration is inherently tied to the physical node (for
+example networking, with bonds, vlans etc. or hardware specific setup such as
+drivers etc.). There is a need for giving the user the possibility to provide
+a node-specific userdata. In this proposal we outline a solution that would
+merge different userdata inputs.
+
+## Motivation
+
+### Goals
+
+* Provide support for user-provided cloud-config per node
+* Validate the user-provided cloud-config (to prevent conflict with CABPK
+  output)
+* Merge the inputs from CABPK and user
+* Support later addition of further configurations (Ignition etc.).
+
+### Non-Goals
+
+ * Support other type of userdata, considering that Cluster-api only supports
+   cloud-init based bootstrap provider (CABPK) we will focus on cloud-init only
+   in this proposal.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+Considering a set of motley physical nodes, all with different networking
+setup, the user will be able to specify a per-node network configuration. Hence
+the BaremetalHost selection can be independent, not requiring to tie
+BaremetalHost and BaremetalMachines through labels to give the networking
+through a script passed to CABPK.
+
+#### Story 2
+
+Considering a CAPI deployment, it would be possible to specify some
+node-specific configuration related to the physical node, offering the user the
+possibility to fully use the deployment features such as scale-in/scale-out
+independently. The deployment could then use physical nodes that are not
+strictly identical.
+
+### Implementation Details/Notes/Constraints
+
+This does not require to modify the current CAPI workflow.
+When selecting a baremetalHost for the BaremetalMachine, the controller would
+merge the userdata from the machine with the user-provided input and the
+load-balancer related input. The output would be written as userdata in the
+BaremetalHost for deployment.
+
+### Risks and Mitigations
+
+There are several types of userData, such as cloud-init or ignition. This
+proposal should not implement a feature for one that would be detrimental or not
+consider another one.
+
+The user could also give some input that directly conflict with the output of
+the bootstrap provider. This needs to be handled carefully.
```

Comment:
One question is how much can we realistically validate this beyond looking for conflicting cloud-config keys (which are actually allowed if you use the cloud-init merge strategies implementation), there could be subtle incompatibility which doesn't explicitly overlap (such as modifying storage or networking in a way that breaks the cluster configuration)?